File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        char ch[256]={0};
        int l=0, ans=0;
        for(int r=0;r<s.size(); r++) {
            ch[s[r]]++;
            while(l<r && ch[s[r]]>1) {
                ch[s[l]]--;
                l++;
            }
            ans = max(ans, r-l+1);
        }
        return ans;
    }
};

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-1.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [-1] * (n+1)
        def solve(idx: int, num: int) -> int:
            if idx == n:
                return 0
            best = 0
            if dp[idx] != -1:
                return dp[idx]
            for i in range(idx, n):
                if nums[i] > num:
                    length = 1 + solve(i+1, nums[i])
                    best = max(best, length)
            dp[idx] = best
            return dp[idx]
                    
        maxLen = 0
        for i, num in enumerate(nums):
            length = 1+ solve(i+1, num)
            maxLen = max(maxLen, length)
        return maxLen